# 4 Pythonによるデータ加工処理の基礎（Pandas）

実際にデータ分析を行う際には、データを把握し、整える前処理を行う必要があります。この章では、Pandasというデータ分析でよく使うライブラリについて、基本的な使い方を紹介していきます。この後の章でも使用するライブラリなので、ここでしっかりと基礎を押さえておきましょう。

なお、より詳細について学びたい場合は公式ドキュメント([https://pandas.pydata.org/](https://pandas.pydata.org/))を参照してください。

### 目次

**[4.1 Pandasの概要](#4.1-Pandasの概要)** 
> [4.1.1 ライブラリの読み込み](#4.1.1-ライブラリの読み込み)

**[4.2 Pandasのデータ構造](#4.2-Pandasのデータ構造)** 
> [4.2.1 Seriesオブジェクト](#4.2.1-Seriesオブジェクト)
>
> [4.2.2 DataFrameオブジェクト](#4.2.2-DataFrameオブジェクト)

**[4.3 データの読み込みと対話](#4.3-データの読み込みと対話)** 
> [4.3.1 インターネットなどで配布されている対象データの読み込み](#4.3.1-インターネットなどで配布されている対象データの読み込み)
>
> [4.3.2 データの読み込みと確認](#4.3.2-データの読み込みと確認)

**[4.4 DataFrameの基本](#4.4-DataFrameの基本)** 
> [4.4.1 データの選択と代入](#4.4.1-データの選択と代入)
>
> [4.4.2 特定の条件のデータの抽出](#4.4.2-特定の条件のデータの抽出)
>
> [4.4.3 値のソート](#4.4.3-値のソート)
>
> [4.4.4 nan（null）の判定の基礎](#4.4.4-nan（null）の判定の基礎)
>
> [4.4.5 データの結合](#4.4.5-データの結合)
>
> [4.4.6 データの削除](#4.4.6-データの削除)
>
> [4.4.7 階層型インデックス](#4.4.7-階層型インデックス)
>
> [4.4.8 データの操作と変換](#4.4.8-データの操作と変換)
>
> [4.4.9 データの集約とグループ演算](#4.4.9-データの集約とグループ演算)

**[4.5 欠損データと異常値の取り扱いの基礎](#4.5-欠損データと異常値の取り扱いの基礎)** 
> [4.5.1 欠損データの扱い方](#4.5.1-欠損データの扱い方)
>
> [4.5.2 異常データの扱い方](#4.5.2-異常データの扱い方)

**[4.6 時系列データの取り扱いの基礎](#4.6-時系列データの取り扱いの基礎)** 
> [4.6.1 時系列データの処理と変換](#4.6.1-時系列データの処理と変換)
>
> [4.6.2 移動平均](#4.6.2-移動平均)

**[4.7 総合問題](#4.7-総合問題)**
> [■総合問題4-1-データ操作](#■総合問題2-1-データ操作)


***

## 4.1 Pandasの概要

Pandasとは、Pythonのデータを数表として扱うことを提供するライブラリです。データフレーム形式でさまざまなデータを加工することができ、データの結合や切り取り、表計算、時系列データの取り扱いなどが可能です。また、簡単なグラフ化も同時にできます。

Pandasは、データサイエンスの分野でよく使われるライブラリで、機械学習を行う前にデータの整形を行う前処理のツールとして使用することが多いです。

### 4.1.1 ライブラリの読み込み

Pandasを利用するためには、ライブラリを読み込む必要があります。読み込むための構文は主に次の2つです。

（1） `import` ライブラリ名 `as` 別名

（2） `from` ライブラリ名 `import` 機能名

(1)は、ライブラリ全体を読み込むために使用します。

(2)は階層化しているライブラリで特定の機能をもつモジュールを読みこむことができます。

目的に応じて使い分けましょう。

In [5]:
import pandas as pd

これは、Pandasというライブラリを `pd` という別名で読み込んでいます。Pandasは一般的に `pd` という別名を与えて読み込みます。Pandasは、機能が階層化されているため、以降では、 `pd.機能名` または `pd.機能名.機能名...`とすることで必要な機能を呼び出すことができます。

In [6]:
from pandas import Series, DataFrame

これは、Pandasの `Series` と `DataFrame` という機能を読み込んでいます。このようにすることで、`pd.Series` または `pd.DataFrame`とする必要があったものを簡単に `Series` または `DataFrame`のように書くことができます。

#### この章で使うライブラリの読み込み


In [7]:
# 以下のライブラリを使うので、あらかじめ読み込んでおいてください
import numpy as np
import numpy.random as random
import pandas as pd
from pandas import Series, DataFrame

***

## 4.2 Pandasのデータ構造

Pandasではよく使われるデータ構造として `Series` と `DataFrame` があります。`Series` は一次元、`DataFrame` は基本的に２次元のデータ列を扱うためのオブジェクトです。

### 4.2.1 Seriesオブジェクト

`Series`オブジェクトは1次元のデータ列です。`Series`は前の章で学習した1次元のNumpyの`array`にラベルがついたようなオブジェクトです。

In [141]:
# 例
series = Series([1,1,2,3,5,8,13])
print(series)

0     1
1     1
2     2
3     3
4     5
5     8
6    13
dtype: int64


`Series`オブジェクトは要素（上右列）を特定するインデックス（上左列）からなります。例では [1,1,2,3,5,8,13] を要素にもつ `Series` オブジェクトを作成しました。`dtype`はデータの型を表します。この例のように値だけを指定した場合、インデックスは先頭から0、1、2…のようになります。

インデックスには任意の数値や文字を指定することができます。

In [142]:
# indexをアルファベットでつける
series_i = Series(
    [1,1,2,3,5,8,13],
    index=['a', 'b', 'c', 'd', 'e', 'f','g'])
print(series_i)

a     1
b     1
c     2
d     3
e     5
f     8
g    13
dtype: int64


要素とインデックスは、それぞれ次のように、`values`プロパティと`index`プロパティを指定することで、別々に取り出すこともできます。`values`プロパティで返された要素はNumpyの`array`形式となります。

In [143]:
print('要素:', series_i.values)
print('インデックス:', series_i.index)

要素: [ 1  1  2  3  5  8 13]
インデックス: Index(['a', 'b', 'c', 'd', 'e', 'f', 'g'], dtype='object')


### 4.2.2 DataFrameオブジェクト

`DataFrame`オブジェクトは基本的に2次元のデータ列です。下記は、`ID`、`City`、`Birth_year`、`Name`の4つの列を持つデータ構造を示した例です。`print`関数で表示すると、そのデータは表形式で表示されます。

In [144]:
data = {'ID':['100','101','102','103','104'],
            'City':['Tokyo','Osaka','Kyoto','Hokkaido','Tokyo'],
            'Birth_year':[1990,1989,1992,1997,1982],
            'Name':['Hiroshi','Akiko','Yuki','Satoru','Steve']}

df = DataFrame(data)

print(df)

    ID      City  Birth_year     Name
0  100     Tokyo        1990  Hiroshi
1  101     Osaka        1989    Akiko
2  102     Kyoto        1992     Yuki
3  103  Hokkaido        1997   Satoru
4  104     Tokyo        1982    Steve


一番左列に表示されている`0, 1, 2, 3, 4`の値は、`Index`の値です。一番上行に表示されている`ID`、`City`、`Birth_year`、`Name`の値は、`Column`の値です。



`DataFrame`オブジェクトも`Series`オブジェクトと同様に、`Index`として文字を指定したりすることができます。

In [145]:
df_i = DataFrame(data,index=['a','b','c','d','e'])
print(df_i)

    ID      City  Birth_year     Name
a  100     Tokyo        1990  Hiroshi
b  101     Osaka        1989    Akiko
c  102     Kyoto        1992     Yuki
d  103  Hokkaido        1997   Satoru
e  104     Tokyo        1982    Steve


要素とインデックスとカラムは、それぞれ次のように、`values`プロパティと`index`プロパティと`columns`プロパティを指定することで、別々に取り出すこともできます。`values`プロパティで返された要素はNumpyのarray形式となります。

In [146]:
print('要素:', df_i.values)
print('インデックス:', df_i.index)
print('カラム:', df_i.columns)

要素: [['100' 'Tokyo' 1990 'Hiroshi']
 ['101' 'Osaka' 1989 'Akiko']
 ['102' 'Kyoto' 1992 'Yuki']
 ['103' 'Hokkaido' 1997 'Satoru']
 ['104' 'Tokyo' 1982 'Steve']]
インデックス: Index(['a', 'b', 'c', 'd', 'e'], dtype='object')
カラム: Index(['ID', 'City', 'Birth_year', 'Name'], dtype='object')


#### Jupyter環境におけるデータ表示

Jupyterでは、データの変数をそのまま次のように記述することで表示することもできます。
この場合、Jupyter環境によって、これが`DataFrame`オブジェクトであることが認識され、罫線などが付いた見やすい表示になります。

In [147]:
df_i

ID      City  Birth_year     Name
a  100     Tokyo        1990  Hiroshi
b  101     Osaka        1989    Akiko
c  102     Kyoto        1992     Yuki
d  103  Hokkaido        1997   Satoru
e  104     Tokyo        1982    Steve

また、大きな`DataFrame`を表示しようとすると一部省略されて表示されることがあります。以下で最大表示列数または最大表示行数を指定することができます。

In [148]:
#最大表示列数の指定
pd.set_option('display.max_columns', 50)
#最大表示行数の指定
pd.set_option('display.max_rows', 50)

***

## 4.3 データの読み込みと対話
データを解析するには、対象のデータをPythonで扱えるように読み込む必要があります。 データはCSV形式データやデータベースとして扱うのが一般的です。またインターネットには、研究用のデータが圧縮されたZIP形式で提供されているものもあります。 まずは、こうしたデータを読み込む方法から習得しましょう。

### 4.3.1 インターネットなどで配布されている対象データの読み込み
ここでは対象のデータが、ZIP形式ファイルとしてWebで公開されており、それをダウンロードして利用するという状況を想定します。ブラウザからあらかじめダウンロードしておくこともできますが、Pythonでは、直接読み込んでデータを保存することもできるため、ここでは、Pythonのプログラムでダウンロードする方法を説明します。

#### サンプルデータのダウンロード

次に、このディレクトリにサンプルデータをダウンロードします。ここでは、カリフォルニア大学アーバイン校（UCI）が提供しているサンプルデータを利用します。 ここではファイルをPythonのプログラムでダウンロードすることにします。

#### ZIPファイルとファイルをダウンロードするためのライブラリ
まずは、ZIPファイルやファイルをダウンロードするためのライブラリをインポートします。ZIPファイルを読み込んだり、Webから直接ダウンロードしたりするには、次のように「`requests`」「`zipfile`」「`io`」の3つのライブラリを使います。

- `requests` … Webのデータを送受信します
- `zipfile` … ZIP形式ファイルを読み書きします
- `io` … ファイルを読み書きします

In [149]:
# webからデータを取得したり、zipファイルを扱うためのライブラリ
import requests, zipfile
from io import StringIO
import io

#### ZIPファイルをダウンロードして展開する
ここで利用するファイルは、次のファイルです。ZIP形式でまとめられています。

http://archive.ics.uci.edu/ml/machine-learning-databases/00356/student.zip

このファイルをダウンロードして展開するには、次のPythonプログラムをJupyter環境のセルに入力して実行します。すると、現在のカレントディレクトリに展開されます。

In [150]:
# データがあるurlの指定
url = 'http://archive.ics.uci.edu/ml/machine-learning-databases/00356/student.zip'

# データをurlから取得する
r = requests.get(url, stream=True)

# zipfileを読み込み展開する
z = zipfile.ZipFile(io.BytesIO(r.content))
z.extractall()

Webからデータをダウンロードするには、`requests.get`を使います。このダウンロードしたデータを、`io.BytesIO`を使ってバイナリストリームとしてZipFileオブジェクトに与え、最後に`extractall()`を実行すると、ダウンロードしたZIP形式データを展開できます。 ダウンロードが終了したら、データがちゃんとダウンロードされ、展開されているかチェックしましょう。次のように`ls`コマンドを実行すると、カレントディレクトリのファイル一覧を表示できます。

In [151]:
ls

answer4.ipynb   data/            student-merge.R  student.txt
chapter4.ipynb  student-mat.csv  student-por.csv


無事に展開されると、「student.txt」「student-mat.csv」「student-merge.R」「student-por.csv」という4つのファイルが配置されます。これらのデータのうち、「student-mat.csv」のデータを使います。

### 4.3.2 データの読み込みと確認

まずは、「student-mat.csv」が、どのようなデータであるのかを観察していきます。

#### データをDataFrameとして読み込む
まずは、対象のデータを読み取り、Pandasの`DataFrame`オブジェクトとして扱います。次のように`pd.read_csv`の引数にファイル名student-mat.csvファイルを記載して実行すると、そのファイルが読み込まれ、`DataFrame`オブジェクトとなります。

In [152]:
student_data_math = pd.read_csv('student-mat.csv')

#### データを確認する
データを読み込んだら、実際のデータの中身を見てみましょう。`head`を使うと、データの先頭から一部をサンプルとして参照できます。括弧のなかに何も指定しない場合は先頭の5行が表示されますが、括弧のなかに行数を指定した場合は、指定した行数だけ表示されます。たとえば、`head(10)`とすれば、10行分表示されます。

In [155]:
student_data_math.head()

school;sex;age;address;famsize;Pstatus;Medu;Fedu;Mjob;Fjob;reason;guardian;traveltime;studytime;failures;schoolsup;famsup;paid;activities;nursery;higher;internet;romantic;famrel;freetime;goout;Dalc;Walc;health;absences;G1;G2;G3
0  GP;"F";18;"U";"GT3";"A";4;4;"at_home";"teacher...                                                                                                                                                                                 
1  GP;"F";17;"U";"GT3";"T";1;1;"at_home";"other";...                                                                                                                                                                                 
2  GP;"F";15;"U";"LE3";"T";1;1;"at_home";"other";...                                                                                                                                                                                 
3  GP;"F";15;"U";"GT3";"T";4;2;"health";"services...                                                                                                                                                                                 
4  GP;"F";16;"U";"GT3";"T";3;3;"other";"other";"h...

#### カンマで区切ってデータを読む
データが入っているのはわかりますが、このままではデータが大変扱いにくいです。よくデータを見てみると、ダウンロードしたデータの区切り文字は「;」（セミコロン）となっています。ほとんどのCSV形式ファイルでは「,」（カンマ）がデータの区切り文字として使われるのが慣例なのですが、ダウンロードしたデータは「;」が区切りであるため、データの区切りを正しく識別できないので、このようにデータがつながってしまうのです。

区切り文字を変えるには、read＿csvのバラメータとして`sep='区切り文字'`を指定します。「;」を区切り文字にするため、次のようにして、データを再度読み込みましょう。

In [156]:
# データの読み込み
# 区切りに";"がついているので注意
student_data_math = pd.read_csv('student-mat.csv', sep=';')
student_data_math.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher   
1     GP   F   17       U     GT3       T     1     1  at_home     other   
2     GP   F   15       U     LE3       T     1     1  at_home     other   
3     GP   F   15       U     GT3       T     4     2   health  services   
4     GP   F   16       U     GT3       T     3     3    other     other   

   reason guardian  traveltime  studytime  failures schoolsup famsup paid  \
0  course   mother           2          2         0       yes     no   no   
1  course   father           1          2         0        no    yes   no   
2   other   mother           1          2         3       yes     no  yes   
3    home   mother           1          3         0        no    yes  yes   
4    home   father           1          2         0        no    yes  yes   

  activities nursery higher internet romantic  famrel  freetime  goout  Dalc  \
0         no     yes    yes       no       no       4         3      4     1   
1         no      no    yes      yes       no       5         3      3     1   
2         no     yes    yes      yes       no       4         3      2     2   
3        yes     yes    yes      yes      yes       3         2      2     1   
4         no     yes    yes       no       no       4         3      2     1   

   Walc  health  absences  G1  G2  G3  
0     1       3         6   5   6   6  
1     1       3         4   5   5   6  
2     3       3        10   7   8  10  
3     1       5         2  15  14  15  
4     2       5         4   6  10  10

データが正しく区切られました。

なお、read_csvの解説を見ると最初から「;」が設定されていることが多いのですが、まだ何も知らない見たこともないデータに対して、区切り文字を「;」にすればよいかどうかは、普通はわかりません。データ分析の実務では、試行錯誤をしながら区切り文字を探すことも多いので、今回は上記のような流れで実施してみました。

なお、このread_csvについては、sep以外にもパラメータがいくつかあり、区切り文字のほか、データ名（アドレス含む）、ヘッダーがあるかないかを指定することもできます。どんなパラメータが設定できるのかは、次のように実行すると確認できます。

In [157]:
?pd.read_csv

#### 4.3.3 データの性質を確認する
先ほど読み込んだデータを見てみると、schoolやageなど学生の属性情報が入っているというのはわかります。しかし、いくつデータがあるのか、どんなデータの種類があるのかまだわかりません。

データの個数や型を確認する
次のように`info`を使うと、すべての変数について、`null`でないデータの個数や変数の型などがわかります。

In [158]:
# すべてのカラムの情報等チェック
student_data_math.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 395 entries, 0 to 394
Data columns (total 33 columns):
school        395 non-null object
sex           395 non-null object
age           395 non-null int64
address       395 non-null object
famsize       395 non-null object
Pstatus       395 non-null object
Medu          395 non-null int64
Fedu          395 non-null int64
Mjob          395 non-null object
Fjob          395 non-null object
reason        395 non-null object
guardian      395 non-null object
traveltime    395 non-null int64
studytime     395 non-null int64
failures      395 non-null int64
schoolsup     395 non-null object
famsup        395 non-null object
paid          395 non-null object
activities    395 non-null object
nursery       395 non-null object
higher        395 non-null object
internet      395 non-null object
romantic      395 non-null object
famrel        395 non-null int64
freetime      395 non-null int64
goout         395 non-null int64
Dalc          395 no

また`describe`を使うことで量的データに関して基本的な統計量を算出することができます。

In [159]:
# 基本統計量をチェック
student_data_math.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

## 4.4 DataFrameの基本

以下では、DataFrameオブジェクトの取り扱いについて基本的なデータ操作から学んでみましょう。

#### 転置

行列の転置のように、行と列を入れ替える場合には、`.T`メソッドを使います。

In [161]:
# 転置
df

ID      City  Birth_year     Name
0  100     Tokyo        1990  Hiroshi
1  101     Osaka        1989    Akiko
2  102     Kyoto        1992     Yuki
3  103  Hokkaido        1997   Satoru
4  104     Tokyo        1982    Steve

### 4.4.1 データの選択と代入
列や行の選択を行う方法は複数存在します。

#### 特定列のみを取り出す

特定の列だけを指定したいときは、データの後にその列名を指定します。複数の列を指定したいときは、それらをリスト形式で指定します。

In [162]:
# 列名の指定（1つの場合）
df['Birth_year']

0    1990
1    1989
2    1992
3    1997
4    1982
Name: Birth_year, dtype: int64

In [163]:
# このように記述することも可能
df.Birth_year

0    1990
1    1989
2    1992
3    1997
4    1982
Name: Birth_year, dtype: int64

In [164]:
# 列名の指定(複数の場合)
df[['ID', 'Birth_year']]

ID  Birth_year
0  100        1990
1  101        1989
2  102        1992
3  103        1997
4  104        1982

参照した列は代入を行うことができます。特に、存在しない列に代入を行うと新しい列が作成されます。

In [165]:
# 新しい列の作成
df['Score'] = np.arange(5)*10
df

ID      City  Birth_year     Name  Score
0  100     Tokyo        1990  Hiroshi      0
1  101     Osaka        1989    Akiko     10
2  102     Kyoto        1992     Yuki     20
3  103  Hokkaido        1997   Satoru     30
4  104     Tokyo        1982    Steve     40

#### 特定行のみを取り出す

特定の行だけを指定したいときは、`:`を用いて行指定をします。このとき、`:`の前後には取得したい行の上行を左に、下行を右に指定します。

In [166]:
# 列名の指定
df[0:3]

ID   City  Birth_year     Name  Score
0  100  Tokyo        1990  Hiroshi      0
1  101  Osaka        1989    Akiko     10
2  102  Kyoto        1992     Yuki     20

`Index`が文字であっても同様に指定することができます。

In [167]:
# 列名の指定
df_i['a':'c']

ID   City  Birth_year     Name
a  100  Tokyo        1990  Hiroshi
b  101  Osaka        1989    Akiko
c  102  Kyoto        1992     Yuki

#### df.locの利用
`df.loc[]`はラベルを指定することでNumpyのように特定の行や列を取り出すことができます。

In [168]:
# 行名の指定（１つの場合）
df.loc[0]

ID                100
City            Tokyo
Birth_year       1990
Name          Hiroshi
Score               0
Name: 0, dtype: object

In [170]:
# 行名の指定（複数の場合）
df.loc[0:3]

ID      City  Birth_year     Name  Score
0  100     Tokyo        1990  Hiroshi      0
1  101     Osaka        1989    Akiko     10
2  102     Kyoto        1992     Yuki     20
3  103  Hokkaido        1997   Satoru     30

In [34]:
# 列名の指定（１つの場合）
df.loc[:, 'Birth_year']

0    1990
1    1989
2    1992
3    1997
4    1982
Name: Birth_year, dtype: int64

In [35]:
# 列名の指定（複数の場合）
df.loc[:, ['ID','Birth_year']]

ID  Birth_year
0  100        1990
1  101        1989
2  102        1992
3  103        1997
4  104        1982

In [36]:
# 0から3行かつ'ID'または'Birth_year'の列を指定
df.loc[0:3, ['ID','Birth_year']]

ID  Birth_year
0  100        1990
1  101        1989
2  102        1992
3  103        1997

#### df.ilocの利用
`df.iloc[]`は，整数`Index`を指定することで`df.loc[]`*と同じ要領で特定の行や列を取り出すことができます。*

In [37]:
# 0から3行かつ'ID'または'Birth_year'の列を指定
df.iloc[0:4, 0:2]

ID      City
0  100     Tokyo
1  101     Osaka
2  102     Kyoto
3  103  Hokkaido

#### df.at, df.iatの利用
`df.at[]`と`df.iat[]`は，単独の要素を取得する時に用います。`df.loc[]`や`df.iloc[]`と同じ要領で特定の要素を取り出すことができます。

In [38]:
# 0行目の'ID'を取得
df.at[0,'ID']

'100'

以上の`df.loc[]`,`df.iloc[]`,`df.at[]`,`df.iat[]`は`=`を用いて値を代入することができます。

In [39]:
df.at[0,'Birth_year'] = 2000
df

ID      City  Birth_year     Name  Score
0  100     Tokyo        2000  Hiroshi      0
1  101     Osaka        1989    Akiko     10
2  102     Kyoto        1992     Yuki     20
3  103  Hokkaido        1997   Satoru     30
4  104     Tokyo        1982    Steve     40

### 4.4.2 特定の条件のデータの抽出

`DataFrame`オブジェクトでは、Numpyのように特定の条件を満たすデータだけを取り出したり、複数のデータを結合したりすることもできます。

次の例は、データのうち、`City`が`Tokyo`のみのデータを抽出する例です。この処理は、`df['City'] == 'Tokyo'`が`True`であるデータをすべて`df`から抽出するもので、フィルターの役割を果たしています。

In [40]:
#　条件（フィルター）
df[df['City'] == 'Tokyo']

ID   City  Birth_year     Name  Score
0  100  Tokyo        2000  Hiroshi      0
4  104  Tokyo        1982    Steve     40

ここで指定している条件である`df['City'] == 'Tokyo'`は、`dtype`が`bool`である`Series`オブジェクトです。

In [41]:
print(df['City'] == 'Tokyo')

0     True
1    False
2    False
3    False
4     True
Name: City, dtype: bool


条件を複数指定したいときは、条件式を`()` でくくります。 論理積 (and) は `&`、論理和 (or) は `|`を用います。

In [42]:
# 'Birth_year'が1990以上かつ1995以下
df[(df['Birth_year'] >= 1990) & (df['Birth_year'] < 1995)]

ID   City  Birth_year  Name  Score
2  102  Kyoto        1992  Yuki     20

次のように`isin(リスト)`を使うこともできます。

In [43]:
#　'City'が'Tokyo'または'Osaka'
df[df['City'].isin(['Tokyo','Osaka'])]

ID   City  Birth_year     Name  Score
0  100  Tokyo        2000  Hiroshi      0
1  101  Osaka        1989    Akiko     10
4  104  Tokyo        1982    Steve     40

### 4.4.3 値のソート

`Series`オブジェクトや`DataFrame`オブジェクトのデータは、ソートすることもできます。値だけではなく、`Index`をベースにソートできます。

まずはソート対象のサンプルデータを次のように定義します。ソートの効果がわかりやすくなるよう、わざとデータを適当な順で並べてあります。

In [44]:
# データの準備
data = {'ID':['100','101','102','103','104'],
            'City':['Tokyo','Osaka','Kyoto','Hokkaido','Tokyo'],
            'Birth_year':[1990,1989,1992,1997,1982],
            'Name':['Hiroshi','Akiko','Yuki','Satoru','Steve']}

df_i2 = DataFrame(data,index=['e','b','a','d','c'])
df_i2

ID      City  Birth_year     Name
e  100     Tokyo        1990  Hiroshi
b  101     Osaka        1989    Akiko
a  102     Kyoto        1992     Yuki
d  103  Hokkaido        1997   Satoru
c  104     Tokyo        1982    Steve

`Index`でソートするには、次のように`sort_index`メソッドを実行します。

In [45]:
# indexによるソート
df_i2.sort_index()

ID      City  Birth_year     Name
a  102     Kyoto        1992     Yuki
b  101     Osaka        1989    Akiko
c  104     Tokyo        1982    Steve
d  103  Hokkaido        1997   Satoru
e  100     Tokyo        1990  Hiroshi

値でソートする場合には、次のように`sort_values`メソッドを使います。また、デフォルトは昇順ですが`ascending`を指定することによって降順にすることも可能です。

In [46]:
# 値によるソート
df_i2.sort_values(by='Birth_year', ascending=False)

ID      City  Birth_year     Name
d  103  Hokkaido        1997   Satoru
a  102     Kyoto        1992     Yuki
e  100     Tokyo        1990  Hiroshi
b  101     Osaka        1989    Akiko
c  104     Tokyo        1982    Steve

### 4.4.4 nan（null）の判定

データ分析ではデータが欠損しており、該当のデータが存在しないことがあります。それらをそのまま計算すると、平均などを求めたときに正しい値が得られないので、除外するなどの操作が必要です。欠損値などのデータは`nan`という特別な値で格納されるので、その扱いについて補足します。

#### 条件に合致したデータの比較

まずは`nan`の話ではなく、ふつうに条件検索する例から説明します。
次の例は、`df_i2`の全要素を対象に、`Tokyo`という文字列があるかどうかを`isin`で調べる例です。その結果は、それぞれのセルに`True`か`False`が返されます。入っていれば（条件を満たしていれば）`True`、入っていなければ（条件を満たしていなければ）`False`が設定されます。この操作が、条件に合致するデータを探すときの基本です。

In [47]:
# 値があるかどうかの確認
df_i2.isin(['Tokyo'])

ID   City  Birth_year   Name
e  False   True       False  False
b  False  False       False  False
a  False  False       False  False
d  False  False       False  False
c  False   True       False  False

#### nanとnullの例

次の例は、`Name`列の値をわざと`nan`に設定した例です。`nan`かどうかを判定するには`isnull`メソッドを使います。

In [48]:
#　欠損値の取り扱い
# name をすべてnanにする
df_i2['Name'] = np.nan
df_i2.isnull()

ID   City  Birth_year  Name
e  False  False       False  True
b  False  False       False  True
a  False  False       False  True
d  False  False       False  True
c  False  False       False  True

そして`nan`であるものの総数を求めるには、`sum`メソッドを呼びます。Nameが5になっているのは、上記の結果でわかるように、Trueが5つあるため、それをカウントしているからです。

In [49]:
# nullを判定し、合計する
df_i2.isnull().sum()

ID            0
City          0
Birth_year    0
Name          5
dtype: int64

### 4.4.5 データの結合
私たちがよく扱うデータとしてはそれぞれの列に重複しない項目名がついており、縦方向に行として各項目をもったデータが追加されていったものが多いです。ここではよく使用する方法について説明して行きます。

データの結合で主に使用するものには`pd.merge` `DataFrame.join` `pd.concat` `DataFrame.append` などがあります。それぞれの詳細な使用方法は公式ドキュメント([https://pandas.pydata.org/docs/user_guide/merging.html](https://pandas.pydata.org/docs/user_guide/merging.html))を参照してください。

次に提示する`data1`（以下、データ1）と`data2`（以下、データ2）の2つのデータを使います。

In [50]:
# データ1の準備
data1 = {
    'id': ['100', '101', '102', '103', '104', '106', '108', '110', '111',' 113'],
    'city': ['Tokyo', 'Osaka', 'Kyoto', 'Hokkaido', 'Tokyo', 'Tokyo', 'Osaka', 'Kyoto', 'Hokkaido', 'Tokyo'],
    'birth_year': [1990, 1989, 1992, 1997, 1982, 1991, 1988, 1990, 1995, 1981],
    'name': ['Hiroshi', 'Akiko', 'Yuki', 'Satoru', 'Steeve', 'Mituru', 'Aoi', 'Tarou', 'Suguru','Mitsuo']
}
df1 = DataFrame(data1)
df1

id      city  birth_year     name
0   100     Tokyo        1990  Hiroshi
1   101     Osaka        1989    Akiko
2   102     Kyoto        1992     Yuki
3   103  Hokkaido        1997   Satoru
4   104     Tokyo        1982   Steeve
5   106     Tokyo        1991   Mituru
6   108     Osaka        1988      Aoi
7   110     Kyoto        1990    Tarou
8   111  Hokkaido        1995   Suguru
9   113     Tokyo        1981   Mitsuo

In [51]:
# データ2の準備
data2 = {
    'id': ['100', '101', '102', '105', '107'],
    'math': [50, 43, 33, 76, 98],
    'english': [90, 30, 20, 50, 30],
    'sex': ['M','F','F','M','M'], 
    'index_num': [0, 1, 2, 3, 4]
}
df2 = DataFrame(data2)
df2

id  math  english sex  index_num
0  100    50       90   M          0
1  101    43       30   F          1
2  102    33       20   F          2
3  105    76       50   M          3
4  107    98       30   M          4

ここでは、この2つのデータを結合する方法を見ていきましょう。 データ1とデータ2を結合する方法は、次の4パターンが考えられます。

①内部結合（INNER JOIN） 両方にキーがあるときに結合します。

②全結合（FULL JOIN） どちらかにキーがあるときに結合します。

③左外部結合（LEFT JOIN） 左側にあるデータのキーがある時に結合します。

④右外部結合（RIGHT JOIN） 右側にあるデータのキーがある時に結合します。

![](https://mathwords.net/wp-content/uploads/2017/06/tablejoin1.png)

引用URL:https://mathwords.net/wp-content/uploads/2017/06/tablejoin1.png

#### `pd.merge`を使用
`pd.merge`は項目名のうち一つをキーに指定して結合するときによく使用します。

##### 内部結合
mergeメソッドの結合方法のデフォルトは内部結合です。`on`引数でキーを指定することができますが、指定しなかった場合は両方の`DataFrame`に同じ名前で含まれるカラム`id`をキーとして内部結合します。なお、`key`は複数指定することもでき、その場合はカラム名をリストで渡します。

In [52]:
# データのマージ（内部結合。キーは自動的に認識される）
pd.merge(df1, df2)

id   city  birth_year     name  math  english sex  index_num
0  100  Tokyo        1990  Hiroshi    50       90   M          0
1  101  Osaka        1989    Akiko    43       30   F          1
2  102  Kyoto        1992     Yuki    33       20   F          2

In [53]:
# データのマージ（内部結合。onで明示的に指定可能）
pd.merge(df1, df2, on = 'id')

id   city  birth_year     name  math  english sex  index_num
0  100  Tokyo        1990  Hiroshi    50       90   M          0
1  101  Osaka        1989    Akiko    43       30   F          1
2  102  Kyoto        1992     Yuki    33       20   F          2

`id`の値が両方の`Dataframe`オブジェクトに存在するものみが表示されました。

なお、`left_on`や`right_on`引数を使うと、キーを個別に指定できます。さらに，`left_index`や`right_index`引数を用いると`Index`で指定して結合できます。次の例は、左側のデータの`Index`と、右側のデータのindex_numカラムをキーとして指定するものです。

In [54]:
# Indexとindex_num によるマージ
pd.merge(df1, df2, left_index = True, right_on = 'index_num')

id_x      city  birth_year     name id_y  math  english sex  index_num
0  100     Tokyo        1990  Hiroshi  100    50       90   M          0
1  101     Osaka        1989    Akiko  101    43       30   F          1
2  102     Kyoto        1992     Yuki  102    33       20   F          2
3  103  Hokkaido        1997   Satoru  105    76       50   M          3
4  104     Tokyo        1982   Steeve  107    98       30   M          4

ここで重複のある`id`は左側のデータは`id_x`右側のデータは`id_y`と変わっています。この値は`suffixes`を指定することで変更することができます

In [55]:
# suffixesを指定
pd.merge(df1, df2, left_index = True, right_on = 'index_num', suffixes=('_1','_2'))

id_1      city  birth_year     name id_2  math  english sex  index_num
0  100     Tokyo        1990  Hiroshi  100    50       90   M          0
1  101     Osaka        1989    Akiko  101    43       30   F          1
2  102     Kyoto        1992     Yuki  102    33       20   F          2
3  103  Hokkaido        1997   Satoru  105    76       50   M          3
4  104     Tokyo        1982   Steeve  107    98       30   M          4

##### 全結合
次の例は、どちらのデータにも存在するデータで結合しています。これが全結合です。全結合では`how`引数に`outer`を指定します。結合する値がない場合は、`NaN`になります。

In [56]:
# データのマージ（全結合）
pd.merge(df1, df2, how = 'outer')

id      city  birth_year     name  math  english  sex  index_num
0    100     Tokyo      1990.0  Hiroshi  50.0     90.0    M        0.0
1    101     Osaka      1989.0    Akiko  43.0     30.0    F        1.0
2    102     Kyoto      1992.0     Yuki  33.0     20.0    F        2.0
3    103  Hokkaido      1997.0   Satoru   NaN      NaN  NaN        NaN
4    104     Tokyo      1982.0   Steeve   NaN      NaN  NaN        NaN
5    106     Tokyo      1991.0   Mituru   NaN      NaN  NaN        NaN
6    108     Osaka      1988.0      Aoi   NaN      NaN  NaN        NaN
7    110     Kyoto      1990.0    Tarou   NaN      NaN  NaN        NaN
8    111  Hokkaido      1995.0   Suguru   NaN      NaN  NaN        NaN
9    113     Tokyo      1981.0   Mitsuo   NaN      NaN  NaN        NaN
10   105       NaN         NaN      NaN  76.0     50.0    M        3.0
11   107       NaN         NaN      NaN  98.0     30.0    M        4.0

##### 左外部結合
左外部結合は`how`引数に`left`を指定します。次の例は、左側のテーブル（ひとつめの引数）に合わせて、`Dataframe`オブジェクトのデータを結合するものです。左側に対応するデータが右（ふたつめの引数）にない場合は、`NaN`になります。

In [57]:
# データのマージ（left）
pd.merge(df1, df2, how = 'left')

id      city  birth_year     name  math  english  sex  index_num
0   100     Tokyo        1990  Hiroshi  50.0     90.0    M        0.0
1   101     Osaka        1989    Akiko  43.0     30.0    F        1.0
2   102     Kyoto        1992     Yuki  33.0     20.0    F        2.0
3   103  Hokkaido        1997   Satoru   NaN      NaN  NaN        NaN
4   104     Tokyo        1982   Steeve   NaN      NaN  NaN        NaN
5   106     Tokyo        1991   Mituru   NaN      NaN  NaN        NaN
6   108     Osaka        1988      Aoi   NaN      NaN  NaN        NaN
7   110     Kyoto        1990    Tarou   NaN      NaN  NaN        NaN
8   111  Hokkaido        1995   Suguru   NaN      NaN  NaN        NaN
9   113     Tokyo        1981   Mitsuo   NaN      NaN  NaN        NaN

#### `DataFrame.join`を使用
`DataFrame.join`は`index`をキーに指定する場合に使用します。使用方法は`pd.merge`に似ています。詳細な使用方法は公式ドキュメント([https://pandas.pydata.org/docs/user_guide/merging.html](https://pandas.pydata.org/docs/user_guide/merging.html))を参照してください。

#### `pd.concat`を使用
`pd.concat`は、データ同士をキーを指定せずに結合する時に使用します。

##### 縦に結合
デフォルトでは縦に単純に接続します。`pd.concat`はよく、同じカラムをもつデータを縦方向に積み上げるときに用います。

In [58]:
# データ3の準備
data3 = {
    'id': ['117', '118', '119', '120', '125'],
    'city': ['Chiba', 'Kanagawa', 'Tokyo', 'Fukuoka', 'Okinawa'],
    'birth_year': [1990, 1989, 1992, 1997, 1982],
    'name': ['Suguru', 'Kouichi', 'Satochi', 'Yukie', 'Akari']
}
df3 = DataFrame(data3)
df3

id      city  birth_year     name
0  117     Chiba        1990   Suguru
1  118  Kanagawa        1989  Kouichi
2  119     Tokyo        1992  Satochi
3  120   Fukuoka        1997    Yukie
4  125   Okinawa        1982    Akari

In [59]:
# concat 縦結合
pd.concat([df1,df3])

id      city  birth_year     name
0   100     Tokyo        1990  Hiroshi
1   101     Osaka        1989    Akiko
2   102     Kyoto        1992     Yuki
3   103  Hokkaido        1997   Satoru
4   104     Tokyo        1982   Steeve
5   106     Tokyo        1991   Mituru
6   108     Osaka        1988      Aoi
7   110     Kyoto        1990    Tarou
8   111  Hokkaido        1995   Suguru
9   113     Tokyo        1981   Mitsuo
0   117     Chiba        1990   Suguru
1   118  Kanagawa        1989  Kouichi
2   119     Tokyo        1992  Satochi
3   120   Fukuoka        1997    Yukie
4   125   Okinawa        1982    Akari

上では`index`がそのまま結合されていましたが、`ignore_index=True`とすることで、上から順番に`index`を付け直すことが可能です。

In [60]:
pd.concat([df1, df3], ignore_index=True)

id      city  birth_year     name
0    100     Tokyo        1990  Hiroshi
1    101     Osaka        1989    Akiko
2    102     Kyoto        1992     Yuki
3    103  Hokkaido        1997   Satoru
4    104     Tokyo        1982   Steeve
5    106     Tokyo        1991   Mituru
6    108     Osaka        1988      Aoi
7    110     Kyoto        1990    Tarou
8    111  Hokkaido        1995   Suguru
9    113     Tokyo        1981   Mitsuo
10   117     Chiba        1990   Suguru
11   118  Kanagawa        1989  Kouichi
12   119     Tokyo        1992  Satochi
13   120   Fukuoka        1997    Yukie
14   125   Okinawa        1982    Akari

`pd.concat`で縦に結合する場合カラム名に紐付けられて接合が行われるます。カラム名が異なる場合、結合する値がない場合には`Nan`が入ります。`sort`引数は`True`とすることで`Column`が並べ替えられますが、`False`とすることで並べ替えられずに結合することができます。

In [61]:
# sort = Trueの場合
pd.concat([df1, df2], sort=True)

birth_year      city  english    id  index_num  math     name  sex
0      1990.0     Tokyo      NaN   100        NaN   NaN  Hiroshi  NaN
1      1989.0     Osaka      NaN   101        NaN   NaN    Akiko  NaN
2      1992.0     Kyoto      NaN   102        NaN   NaN     Yuki  NaN
3      1997.0  Hokkaido      NaN   103        NaN   NaN   Satoru  NaN
4      1982.0     Tokyo      NaN   104        NaN   NaN   Steeve  NaN
5      1991.0     Tokyo      NaN   106        NaN   NaN   Mituru  NaN
6      1988.0     Osaka      NaN   108        NaN   NaN      Aoi  NaN
7      1990.0     Kyoto      NaN   110        NaN   NaN    Tarou  NaN
8      1995.0  Hokkaido      NaN   111        NaN   NaN   Suguru  NaN
9      1981.0     Tokyo      NaN   113        NaN   NaN   Mitsuo  NaN
0         NaN       NaN     90.0   100        0.0  50.0      NaN    M
1         NaN       NaN     30.0   101        1.0  43.0      NaN    F
2         NaN       NaN     20.0   102        2.0  33.0      NaN    F
3         NaN       NaN     50.0   105        3.0  76.0      NaN    M
4         NaN       NaN     30.0   107        4.0  98.0      NaN    M

##### 横に結合
横に結合するためには`axis`引数に軸を指定します。「`axis=0`が行」「`axis=1`が列」です。そのため、ここでは`axis=1`に指定する必要があります。この場合、`index`で紐付けられて、`column`はそのまま結合されます。なお、`axis=0`と指定すると縦に結合します。この`axis`引数は他の場面でも使うので、覚えておいてください。

In [62]:
pd.concat([df1, df2], axis=1)

id      city  birth_year     name   id  math  english  sex  index_num
0   100     Tokyo        1990  Hiroshi  100  50.0     90.0    M        0.0
1   101     Osaka        1989    Akiko  101  43.0     30.0    F        1.0
2   102     Kyoto        1992     Yuki  102  33.0     20.0    F        2.0
3   103  Hokkaido        1997   Satoru  105  76.0     50.0    M        3.0
4   104     Tokyo        1982   Steeve  107  98.0     30.0    M        4.0
5   106     Tokyo        1991   Mituru  NaN   NaN      NaN  NaN        NaN
6   108     Osaka        1988      Aoi  NaN   NaN      NaN  NaN        NaN
7   110     Kyoto        1990    Tarou  NaN   NaN      NaN  NaN        NaN
8   111  Hokkaido        1995   Suguru  NaN   NaN      NaN  NaN        NaN
9   113     Tokyo        1981   Mitsuo  NaN   NaN      NaN  NaN        NaN

#### <練習問題 4-1>

下記のデータテーブルに対して、`city`がTokyoであるか`birth_year`が1990より大きいデータを出力してみましょう。

In [63]:
# データ4の準備
data4 = {
    'id': ['0', '1', '2', '3', '4', '6', '8', '11', '12', '13'],
    'city': ['Tokyo', 'Osaka', 'Kyoto', 'Hokkaido', 'Tokyo', 'Tokyo', 'Osaka', 'Kyoto', 'Hokkaido', 'Tokyo'],
    'birth_year': [1990, 1989, 1992, 1997, 1982, 1991, 1988, 1990, 1995, 1981],
    'name': ['Hiroshi', 'Akiko', 'Yuki', 'Satoru', 'Steeve', 'Mituru', 'Aoi', 'Tarou', 'Suguru', 'Mitsuo']
}
df4 = DataFrame(data4)
df4

id      city  birth_year     name
0   0     Tokyo        1990  Hiroshi
1   1     Osaka        1989    Akiko
2   2     Kyoto        1992     Yuki
3   3  Hokkaido        1997   Satoru
4   4     Tokyo        1982   Steeve
5   6     Tokyo        1991   Mituru
6   8     Osaka        1988      Aoi
7  11     Kyoto        1990    Tarou
8  12  Hokkaido        1995   Suguru
9  13     Tokyo        1981   Mitsuo

In [64]:
# write me!

#### <練習問題  4-2>

下記の2つのデータテーブルに対して、内部結合してみましょう。

In [65]:
# データ4の準備
data4 = {
    'id': ['0', '1', '2', '3', '4', '6', '8', '11', '12', '13'],
    'city': ['Tokyo', 'Osaka', 'Kyoto', 'Hokkaido', 'Tokyo', 'Tokyo', 'Osaka', 'Kyoto', 'Hokkaido', 'Tokyo'],
    'birth_year': [1990, 1989, 1992, 1997, 1982, 1991, 1988, 1990, 1995, 1981],
    'name': ['Hiroshi', 'Akiko', 'Yuki', 'Satoru', 'Steeve', 'Mituru', 'Aoi', 'Tarou', 'Suguru', 'Mitsuo']
}
df4 = DataFrame(data4)
df4

id      city  birth_year     name
0   0     Tokyo        1990  Hiroshi
1   1     Osaka        1989    Akiko
2   2     Kyoto        1992     Yuki
3   3  Hokkaido        1997   Satoru
4   4     Tokyo        1982   Steeve
5   6     Tokyo        1991   Mituru
6   8     Osaka        1988      Aoi
7  11     Kyoto        1990    Tarou
8  12  Hokkaido        1995   Suguru
9  13     Tokyo        1981   Mitsuo

In [66]:
# データ5の準備
data5 = {
    'id': ['0', '1', '3', '6', '8'],
    'math' : [20, 30, 50, 70, 90],
    'english': [30, 50, 50, 70, 20],
    'sex': ['M', 'F', 'F', 'M', 'M'],
    'index_num': [0, 1, 2, 3, 4]
}
df5 = DataFrame(data5)
df5

id  math  english sex  index_num
0  0    20       30   M          0
1  1    30       50   F          1
2  3    50       50   F          2
3  6    70       70   M          3
4  8    90       20   M          4

In [67]:
# write me!

#### <練習問題 4-3>
<練習問題 4-2>のデータを使って、`df4`に対して、以下のデータを縦結合してみましょう。

In [68]:
# データの準備
data6 = {
    'id': ['70', '80', '90', '120', '150'],
    'city': ['Chiba', 'Kanagawa', 'Tokyo', 'Fukuoka', 'Okinawa'],
    'birth_year': [1980, 1999, 1995, 1994, 1994],
    'name': ['Suguru', 'Kouichi', 'Satochi', 'Yukie', 'Akari']
}
df6 = DataFrame(data6)
df6

id      city  birth_year     name
0   70     Chiba        1980   Suguru
1   80  Kanagawa        1999  Kouichi
2   90     Tokyo        1995  Satochi
3  120   Fukuoka        1994    Yukie
4  150   Okinawa        1994    Akari

In [69]:
# write me!

### 4.4.6 データの削除
`DataFrame`オブジェクトで列や行を削除するためには主に`pd.drop`を使います。

#### 列や行の削除

ある特定の列や行を削除するには`drop`メソッドを実行します。`axis`引数に軸を指定します。

・行削除の場合：1つ目の引数に削除したい行の`Index`をリストとして指定します。`axis=0`を指定します。
 
・列の削除の場合：1つ目の引数に削除したい列名をリストとして指定します。`axis=1`を指定します。
 
 次の例は、`Birth_year`列を削除する例です。

In [70]:
# データの列の削除
df.drop(['Birth_year'], axis = 1)

ID      City     Name  Score
0  100     Tokyo  Hiroshi      0
1  101     Osaka    Akiko     10
2  102     Kyoto     Yuki     20
3  103  Hokkaido   Satoru     30
4  104     Tokyo    Steve     40

なお、上記で列を削除しても元のデータの列が削除されたわけではないので、注意しましょう。置き換えたい場合は、あらためて`df = df.drop(['Birth_year'],axis=1)`のように設定します。もしくは、オプションの`inplace=True`を指定すると、元のデータを置き換えることもできます。

また、`axis`を用いないで方向を指定することもできます。その場合、行を削除したい時には`index`引数に削除したい`Index`のリストを、列を削除したい場合は`column`引数に削除したい`Column`のリストを指定します。

In [71]:
# データの行の削除
df.drop(index=[0,2])

ID      City  Birth_year    Name  Score
1  101     Osaka        1989   Akiko     10
3  103  Hokkaido        1997  Satoru     30
4  104     Tokyo        1982   Steve     40

ここで、行が削除されたことによって`Index`が連続ではなくなってしまいました。そこで、新しく`Index`を振り直す手段として`DataFrame.reset_index`を用いてみましょう。

In [72]:
df.drop(index=[0,2]).reset_index()

index   ID      City  Birth_year    Name  Score
0      1  101     Osaka        1989   Akiko     10
1      3  103  Hokkaido        1997  Satoru     30
2      4  104     Tokyo        1982   Steve     40

上では元々の`Index`が新しく`column`となっていることがわかります。元々の`Index`を削除する場合には`drop=True`とします。

### 4.4.7 階層型インデックス

データを複数軸で集計したいとき、設定すると便利なのが**階層型インデックス**です。
今までは、1つのインデックスだけを扱いましたが、複数の軸で階層的にインデックスを設定したいこともあります。階層的にインデックスを設定することで、高次元のデータを低次元で扱うことが可能になり、各階層ごとに集計に便利です。

次に示すデータセットは、インデックスを2段構造で設定した例です。インデックスを設定するには、`index`引数にその値を指定します。この例では、1階層目のインデックスとして`a`と`b`、2階層目のインデックスとして`1`と`2`を設定しています。

また、列の側につけるカラムとして、1階層目に`Osaka`、`Tokyo`、`Osaka`、2階層目に`Blue`、`Red`、`Red`を設定しています。


In [73]:
# 3列3行のデータを作成し、インデックスとカラムを設定
hier_df= DataFrame(
    np.arange(9).reshape((3,3)),
    index = [
        ['a','a','b'],
        [1,2,2]
    ], 
    columns = [
        ['Osaka','Tokyo','Osaka'],
        ['Blue','Red','Red']
    ]
)
hier_df

Osaka Tokyo Osaka
     Blue   Red   Red
a 1     0     1     2
  2     3     4     5
b 2     6     7     8

これらのインデックスやカラムには、名前をつけることもできます。

In [74]:
# indexに名前を付ける
hier_df.index.names =['key1','key2']
# カラムに名前を付ける
hier_df.columns.names =['city','color']
hier_df

city      Osaka Tokyo Osaka
color      Blue   Red   Red
key1 key2                  
a    1        0     1     2
     2        3     4     5
b    2        6     7     8

#### カラムの絞り込み

ここでたとえば、カラムの`city`が`Osaka`のデータだけを見たいとしましょう。次のようにすると、グループの絞り込みができます。

In [75]:
hier_df['Osaka']

color      Blue  Red
key1 key2           
a    1        0    2
     2        3    5
b    2        6    8

#### インデックスを軸にした集計

次はインデックスを軸にした集計の例です。以下の例は、`key2`を軸に合計を計算する例です。

In [76]:
# 階層ごとの要約統計量：行合計
hier_df.sum(level = 'key2', axis = 0)

city  Osaka Tokyo Osaka
color  Blue   Red   Red
key2                   
1         0     1     2
2         9    11    13

同様にして、`color`を軸に平均を計算する場合は、次のようにします。列方向に平均をとる場合は、`mean`メソッドに対して`axis`引数を1に設定します。

In [77]:
# 列平均
hier_df.mean(level = 'color', axis = 1)

color      Blue  Red
key1 key2           
a    1      0.0  1.5
     2      3.0  4.5
b    2      6.0  7.5

#### インデックスの要素の削除

ある`Index`を削除したい場合は、`drop`メソッドを使います。`drop`メソッドを使うと、`Index`の要素を削除できます。次の例では、`key1`の`b`を削除しています。

In [78]:
hier_df.drop(['b'])

city      Osaka Tokyo Osaka
color      Blue   Red   Red
key1 key2                  
a    1        0     1     2
     2        3     4     5

#### <練習問題 4-4>

次のデータに対して、`Kyoto`の列だけ抜き出してみましょう。

In [79]:
hier_df1 = DataFrame(
    np.arange(12).reshape((3,4)),
    index = [['c','d','d'],[1,2,1]],
    columns = [
        ['Kyoto','Nagoya','Hokkaido','Kyoto'],
        ['Yellow','Yellow','Red','Blue']
    ]
)

hier_df1.index.names = ['key1','key2']
hier_df1.columns.names = ['city','color']
hier_df1

city       Kyoto Nagoya Hokkaido Kyoto
color     Yellow Yellow      Red  Blue
key1 key2                             
c    1         0      1        2     3
d    2         4      5        6     7
     1         8      9       10    11

In [80]:
# write me!

#### <練習問題 4-5>

<練習問題 4-4>のデータに対して、`city`をまとめて列同士の平均値を出してください。

In [81]:
# write me!

#### <練習問題 4-6>

<練習問題 4-4>のデータに対して、`key2`ごとに行の合計値を算出してみましょう。


In [82]:
# write me!

### 4.4.8 データの操作と変換
次に、データの操作と変換（ピボット操作、データの重複があった場合の処理、マッピング、ビン分割など）について扱っていきましょう。

#### ピボット操作
まずは、データのピボット操作について学びます。ピボット操作とは、行を列に、列を行にする操作です。もう一度、これまで使ってきた階層テーブル`hier_df`を例に考えます。

In [83]:
# hier_dfを用意
hier_df= DataFrame(
    np.arange(9).reshape((3, 3)),
    index = [
        ['a', 'a', 'b'],
        [1, 2, 2]
    ],
    columns = [
        ['Osaka', 'Tokyo', 'Osaka'],
        ['Blue','Red','Red']
    ]
)
hier_df

Osaka Tokyo Osaka
     Blue   Red   Red
a 1     0     1     2
  2     3     4     5
b 2     6     7     8

次のように`stack`メソッドを実行すると、列が行に入れ替わった`DataFrame`オブジェクトを再構成できます。

In [84]:
#　ピボット操作で「Blue、Red」の列を行に変更
hier_df.stack()

Osaka  Tokyo
a 1 Blue      0    NaN
    Red       2    1.0
  2 Blue      3    NaN
    Red       5    4.0
b 2 Blue      6    NaN
    Red       8    7.0

`unstack`メソッドを使うと、逆の操作が可能です。

In [85]:
# unstackメソッドで、「Blue、Red」の行を列に変更
hier_df.stack().unstack()

Osaka     Tokyo     
     Blue Red  Blue  Red
a 1     0   2   NaN  1.0
  2     3   5   NaN  4.0
b 2     6   8   NaN  7.0

上記のデータ操作では、列にあったものを行に持ってきたり、行であったものを列に持ってきたりしています。これらのテクニックは、データのモデリング前の処理として使うことも多く便利ですので、ぜひ理解して使えるようしてください。

#### 重複データの除去
次は、重複があるデータの処理です。データ分析をしていると、データに重複があることもありますし、自分で実際に集計等していて重複が混じることもあり、そのチェックをするという意味で重要です。

まず例として、重複があるデータを準備します。

In [86]:
#　重複があるデータ
dupli_data = DataFrame({
        'col1': [1, 1, 2, 3, 4, 4, 6, 6],
        'col2': ['a', 'b', 'b', 'b', 'c', 'c', 'b', 'b']
})
print('・元のデータ')
dupli_data

・元のデータ


col1 col2
0     1    a
1     1    b
2     2    b
3     3    b
4     4    c
5     4    c
6     6    b
7     6    b

重複の判定には`duplicated`メソッドを使います。それぞれの行が確認され、重複があるときは、`True`となります。ただし、重複のあるデータでも1回目では`False`となり、2回目から`True`になります。

In [87]:
#　重複判定
dupli_data.duplicated()

0    False
1    False
2    False
3    False
4    False
5     True
6    False
7     True
dtype: bool

`drop_duplicates`メソッドを使うと、重複したデータを削除した結果のデータが返されます。

In [88]:
#　重複削除
dupli_data.drop_duplicates()

col1 col2
0     1    a
1     1    b
2     2    b
3     3    b
4     4    c
6     6    b

#### マッピング処理
次に、マッピング処理を説明します。共通のキーとなるデータに対して、一方の（参照）テーブルからそのキーに対応するデータを引っ張ってくる機能です。以下は、都道府県名と地域名を対応付けた参照データです。

Tokyo（東京）→Kanto（関東）

Hokkaido（北海道）→Hokkaido（北海道）

Osaka（大阪）→Kansai（関西）

Kyoto（京都）→Kansai（関西）

まず次のように参照データを作ります。

In [89]:
# 参照データ
city_map ={
    'Tokyo': 'Kanto',
    'Hokkaido': 'Hokkaido',
    'Osaka': 'Kansai',
    'Kyoto':'Kansai'
}
city_map

{'Tokyo': 'Kanto',
 'Hokkaido': 'Hokkaido',
 'Osaka': 'Kansai',
 'Kyoto': 'Kansai'}

次の例は、`df1`の`city`カラムをベースとして、上の参照データ`city_map`から対応する地域名データを持ってきて、新しく一番右に`region`というカラムとして追加するものです。

In [90]:
#　参照データを結合
# もし対応するデータがなかったら、NaNになる。
df1['region'] = df1['city'].map(city_map)
df1

id      city  birth_year     name    region
0   100     Tokyo        1990  Hiroshi     Kanto
1   101     Osaka        1989    Akiko    Kansai
2   102     Kyoto        1992     Yuki    Kansai
3   103  Hokkaido        1997   Satoru  Hokkaido
4   104     Tokyo        1982   Steeve     Kanto
5   106     Tokyo        1991   Mituru     Kanto
6   108     Osaka        1988      Aoi    Kansai
7   110     Kyoto        1990    Tarou    Kansai
8   111  Hokkaido        1995   Suguru  Hokkaido
9   113     Tokyo        1981   Mitsuo     Kanto

このように新しい`region`というカラムをつけることで、この単位で集計が可能になります。

##### 無名関数とmapを組み合わせる

次は、無名関数と`map`を使って、カラムの中の一部のデータを取り出す処理をする例です。具体的には、`birth_year`の上3桁を取得します。関数適応やループなどを使って要素を1つ1つ取り出して処理するより便利なので、まとめて処理したい場合は、このようなやり方を検討することをおすすめします。

In [91]:
#　birth_year の上3つの数字・文字を取り出す
df1['up_two_num'] = df1['birth_year'].map(lambda x: str(x)[0:3])
df1

id      city  birth_year     name    region up_two_num
0   100     Tokyo        1990  Hiroshi     Kanto        199
1   101     Osaka        1989    Akiko    Kansai        198
2   102     Kyoto        1992     Yuki    Kansai        199
3   103  Hokkaido        1997   Satoru  Hokkaido        199
4   104     Tokyo        1982   Steeve     Kanto        198
5   106     Tokyo        1991   Mituru     Kanto        199
6   108     Osaka        1988      Aoi    Kansai        198
7   110     Kyoto        1990    Tarou    Kansai        199
8   111  Hokkaido        1995   Suguru  Hokkaido        199
9   113     Tokyo        1981   Mitsuo     Kanto        198

#### ビン分割

最後にビン分割について説明します。これは、ある離散的な範囲にデータを分割して集計したい場合に、便利な機能です。具体的には、上のデータの`birth_year`に対して、5年区切りで集計をしたい場合など、ある特定の分割をして計算をしたいときに使います。

たとえば以下のように、1980、1985、1990、1995、2000のように5年単位でビン分割するためのリストを用意し、Pandasの`cut`関数を使うと、そのように分割できます。`cut`関数では、1つ目の引数に分割するデータ、2つ目の引数に分割する境界値を、それぞれ指定します。

In [92]:
#　分割の粒度
birth_year_bins = [1980, 1985, 1990, 1995, 2000]

# ビン分割の実施
birth_year_cut_data = pd.cut(df1.birth_year, birth_year_bins)
birth_year_cut_data

0    (1985, 1990]
1    (1985, 1990]
2    (1990, 1995]
3    (1995, 2000]
4    (1980, 1985]
5    (1990, 1995]
6    (1985, 1990]
7    (1985, 1990]
8    (1990, 1995]
9    (1980, 1985]
Name: birth_year, dtype: category
Categories (4, interval[int64]): [(1980, 1985] < (1985, 1990] < (1990, 1995] < (1995, 2000]]

なお、上記のプログラムでは、「1980～1985」の区切りの中には1980は含まれませんが、1985は含まれています。つまり、指定した基準は、「～より後で、～以前」という区切り方として使われます。この動作は、cut関数に`left`オプションや`right`オプションを指定することで変更できます。

上記の結果を使って、それぞれの数を集計したい場合は、`value_counts`関数を使います。

In [93]:
# 集計結果
pd.value_counts(birth_year_cut_data)

(1985, 1990]    4
(1990, 1995]    3
(1980, 1985]    2
(1995, 2000]    1
Name: birth_year, dtype: int64

labelsパラメータを指定することで、それぞれのビンに名前をつけることもできます。

In [94]:
# 名前をつける
group_names = ['early1980s', 'late1980s', 'early1990s', 'late1990s']
birth_year_cut_data = pd.cut(df1.birth_year, birth_year_bins, labels = group_names)
pd.value_counts(birth_year_cut_data)

late1980s     4
early1990s    3
early1980s    2
late1990s     1
Name: birth_year, dtype: int64

上記では、ビン分割のリストを用意しましたが、あらかじめ分割数を指定したい場合は、以下のように設定できます。この場合、最大値と最小値の間を等間隔で分割することができます。なお、データによってはきれいに割り切れず小数点以下がでてくるので注意しましょう。

In [95]:
# 数字で分割数指定可能。ここでは2つに分割
pd.cut(df1.birth_year, 2)

0      (1989.0, 1997.0]
1    (1980.984, 1989.0]
2      (1989.0, 1997.0]
3      (1989.0, 1997.0]
4    (1980.984, 1989.0]
5      (1989.0, 1997.0]
6    (1980.984, 1989.0]
7      (1989.0, 1997.0]
8      (1989.0, 1997.0]
9    (1980.984, 1989.0]
Name: birth_year, dtype: category
Categories (2, interval[float64]): [(1980.984, 1989.0] < (1989.0, 1997.0]]

また`qcut`関数を使うと、分位点での分割もできます。`qcut`関数を使うことで、ほぼ同じサイズのビンを作成することができます。

In [96]:
pd.value_counts(pd.qcut(df1.birth_year, 2))

(1980.999, 1990.0]    6
(1990.0, 1997.0]      4
Name: birth_year, dtype: int64

ここでは対象としたデータが、1981、1982、1988、1989、1990、1990、1991、1992、1995、1997と、ちょうど中央値にあたる値が2つあるため、6つと4つで分割されました。

このビン分割、はじめ何に使うのかイメージがわきにくいかもしれませんが、具体的には、顧客の購買金額合計を分けて、それぞれの顧客層（優良顧客など）を分析をしたい場合など、マーケティング分析にも使えます。

#### <練習問題 4-7>

数学の成績を示すデータである「student-mat.csv」を読み込み、年齢（`age`）を2倍にしたカラムを末尾に追加してみましょう。

In [97]:
# 例）
student_data_math = pd.read_csv('student-mat.csv', sep=';')

In [98]:
# write me!

#### <練習問題 4-8>

<練習問題 4-7>と同じデータで、「`absences`」のカラムについて、以下の3つのビンに分けてそれぞれの人数を数えてみましょう。なお、`cut`のデフォルトの挙動は右側が閉区間です。今回は、`cut`関数に対して`right=False`のオプションを指定して、右側を開区間としてください。

In [99]:
#　分割の粒度
absences_bins = [0,1,5,100]

In [100]:
# write me!

#### <練習問題 4-9>

上記と同じデータで、「`absences`」のカラムについて、`qcut`関数を用いて3つのビンに分けてみましょう。

In [101]:
# write me!

### 4.4.9 データの集約とグループ演算

ここでは、あるカラムを軸にして集計する処理を学びます。
さらに`groupby`メソッドを使うと、ある特定の列を軸とした集計ができます。以前使った`df1`データを対象に、集約やグループ演算をしたいと思います。

In [102]:
# データを用意（確認）、ただし、region付き
df1

id      city  birth_year     name    region up_two_num
0   100     Tokyo        1990  Hiroshi     Kanto        199
1   101     Osaka        1989    Akiko    Kansai        198
2   102     Kyoto        1992     Yuki    Kansai        199
3   103  Hokkaido        1997   Satoru  Hokkaido        199
4   104     Tokyo        1982   Steeve     Kanto        198
5   106     Tokyo        1991   Mituru     Kanto        199
6   108     Osaka        1988      Aoi    Kansai        198
7   110     Kyoto        1990    Tarou    Kansai        199
8   111  Hokkaido        1995   Suguru  Hokkaido        199
9   113     Tokyo        1981   Mitsuo     Kanto        198

以下のように`groupby`メソッドでグループ化してから`size`メソッドを使うと、それぞれの`city`の値がいくつかあるのかを計算できます。

In [103]:
# サイズ情報
df1.groupby('city').size()

city
Hokkaido    2
Kyoto       2
Osaka       2
Tokyo       4
dtype: int64

次は、`city`を軸として、`birth_year`の平均値を算出する例です。

In [104]:
# cityを軸に、birth_yearの平均値を求める
df1.groupby('city')['birth_year'].mean()

city
Hokkaido    1996.0
Kyoto       1991.0
Osaka       1988.5
Tokyo       1986.0
Name: birth_year, dtype: float64

軸は複数設定することもできます。たとえば、region、cityを2軸として、birth_yearの平均値を求めると、次のようになります。

In [105]:
df1.groupby(['region', 'city'])['birth_year'].mean()

region    city    
Hokkaido  Hokkaido    1996.0
Kansai    Kyoto       1991.0
          Osaka       1988.5
Kanto     Tokyo       1986.0
Name: birth_year, dtype: float64

なお、`groupby`メソッドに`as_index = False`パラメータを設定すると、インデックスが設定されなくなります。そのままテーブルとして扱いたいときに便利です。

In [106]:
df1.groupby(['region', 'city'], as_index = False)['birth_year'].mean()

region      city  birth_year
0  Hokkaido  Hokkaido      1996.0
1    Kansai     Kyoto      1991.0
2    Kansai     Osaka      1988.5
3     Kanto     Tokyo      1986.0

他にも`groupby`メソッドには、イテレータという、反復的に値を取り出す機能があり、次のように、結果の要素をPythonの`for`などでループ処理できて便利です。

以下の例は、groupはregionの名前取り出し、subdfはそのregionのみの行をすべて抽出するというものです。

In [107]:
for group, subdf in df1.groupby('region'):
    print('==========================================================')
    print('Region Name:{0}'.format(group))
    print(subdf)

Region Name:Hokkaido
    id      city  birth_year    name    region up_two_num
3  103  Hokkaido        1997  Satoru  Hokkaido        199
8  111  Hokkaido        1995  Suguru  Hokkaido        199
Region Name:Kansai
    id   city  birth_year   name  region up_two_num
1  101  Osaka        1989  Akiko  Kansai        198
2  102  Kyoto        1992   Yuki  Kansai        199
6  108  Osaka        1988    Aoi  Kansai        198
7  110  Kyoto        1990  Tarou  Kansai        199
Region Name:Kanto
     id   city  birth_year     name region up_two_num
0   100  Tokyo        1990  Hiroshi  Kanto        199
4   104  Tokyo        1982   Steeve  Kanto        198
5   106  Tokyo        1991   Mituru  Kanto        199
9   113  Tokyo        1981   Mitsuo  Kanto        198


データに対して、複数の計算をまとめて行いたいときには、`agg`メソッドを使うと便利です。`agg`メソッドの引数には、実行したい関数名のリストを渡します。以下は、カウント、平均、最大、最小を計算する例です。

なお、以下の例では、student-mat.csvを使って計算しています。このデータがあるファイルディレクトリに移動して、データを読み込んで実行してください。

In [108]:
student_data_math = pd.read_csv('student-mat.csv', sep = ';')

In [109]:
# 列に複数の関数を適応
functions = ['count','mean','max','min']
grouped_student_math_data1 = student_data_math.groupby(['sex','address'])
grouped_student_math_data1['age','G1'].agg(functions)

age                       G1                   
            count       mean max min count       mean max min
sex address                                                  
F   R          44  16.977273  19  15    44  10.295455  19   6
    U         164  16.664634  20  15   164  10.707317  18   4
M   R          44  17.113636  21  15    44  10.659091  18   3
    U         143  16.517483  22  15   143  11.405594  19   5

#### <練習問題 4-10>

「student-mat.csv」を使って、Pandasの集計処理をしてみましょう。まずは、学校（`school`）を軸にして、`G1`の平均点をそれぞれ求めてみましょう。

In [110]:
# write me!

#### <練習問題 4-11>

「student-mat.csv」を使って、学校（`school`）と性別（`sex`）を軸にして、`G1`、`G2`、`G3`の平均点をそれぞれ求めてみましょう。

In [111]:
# write me!

#### <練習問題 4-12>

「student-mat.csv」を使って、学校（`school`）と性別（`sex`）を軸にして、`G1`、`G2`、`G3`の最大値、最小値をまとめて算出してみましょう。

In [112]:
# write me!

## 4.5 欠損データと異常値の取り扱いの基礎
ゴール：欠損データと異常値に対する基本的な対処方法を知る

データを扱っていると必ずといっていいほど、欠損しているデータや異常値データの存在があります。この節では、基礎の基礎レベルで欠損データや異常データについての判定や扱い方について学ぶことにします。もっと深く学びたい方は、ぜひ参考文献「A-12」を読んでください。

### 4.5.1 欠損データの扱い方
まずは、欠損データの取り扱いについてです。データの欠損は、入力忘れ、無回答、システム上の問題などさまざまな要因があります。「ない」データについては、無視をするのがいいのか、除外をするのがいいのか、もっともらしい値を入れるのがいいのか、それが問題です。アプローチによっては、大きなバイアスのある結果を与え、誤った意思決定につながり、大きな損失につながる可能性もあります。慎重に扱っていきましょう。

この節では、次のようなデータをサンプルとして扱います。値を`NaN`（`NA`）にした部分が欠損データであるとして、以下、説明を続けます。

In [113]:
# データの準備
import numpy as np
from numpy import nan as NA
import pandas as pd


df = pd.DataFrame(np.random.rand(10, 4))

# NAにする
df.iloc[1,0] = NA
df.iloc[2:3,2] = NA
df.iloc[5:,3] = NA

In [114]:
df

0         1         2         3
0  0.409686  0.127332  0.232334  0.472368
1       NaN  0.308719  0.099518  0.838625
2  0.735781  0.911987       NaN  0.982552
3  0.333130  0.581156  0.911326  0.673805
4  0.900381  0.254568  0.969245  0.841306
5  0.697316  0.653956  0.577141       NaN
6  0.212289  0.870528  0.917358       NaN
7  0.574902  0.365432  0.778180       NaN
8  0.802476  0.911588  0.448293       NaN
9  0.752408  0.741778  0.980731       NaN

以下では、この擬似的な欠損データに対して、削除や0や直前の数字、平均値等で穴埋めをしていきます。今回は、これらの単純な方法のみ紹介しますが、他の方法として、最尤推定法で推定したり、回帰代入やScipyで実施したスプライン補間などもあります。注意が必要なのは、これらの方法がバイアスを生む可能性があることです。ここで紹介する方法がベストであるとはいえません。深く学びたい方はぜひ参考文献「A-12」などを読んで、欠損データを埋める方法への理解を深めてください。

#### リストワイズ削除
NaNがある行をすべて取り除くには、`dropna`メソッドを使います。これを**リストワイズ削除**といいます。
以下は、先ほどのデータにおいて、`dropna`メソッドを適用し、すべてのカラムにデータがある行だけを抽出したものです。`NaN`がある行は除外されます。

In [115]:
df.dropna()

0         1         2         3
0  0.409686  0.127332  0.232334  0.472368
3  0.333130  0.581156  0.911326  0.673805
4  0.900381  0.254568  0.969245  0.841306

#### ペアワイズ削除
この結果からわかるように、リストワイズ削除では元々10行あったデータが極端に少なくなって、データが全く使えないという状況が考えられます。このとき、欠損している列のデータを無視して、利用可能なデータのみ（例：列の0番目と1番目のみ存在）を使う方法があります。これを**ペアワイズ削除**といいます。ペアワイズ削除では、使いたい列を取り出してから`dropna`メソッドを適用します。

In [116]:
df[[0,1]].dropna()

0         1
0  0.409686  0.127332
2  0.735781  0.911987
3  0.333130  0.581156
4  0.900381  0.254568
5  0.697316  0.653956
6  0.212289  0.870528
7  0.574902  0.365432
8  0.802476  0.911588
9  0.752408  0.741778

#### fillnaで埋める
他の処理として、`fillna`（値）で、`NaN`になっている箇所をある値で埋める方法もあります。たとえば`NaN`を0として扱うケースです。次のように`fillna(0)`とすると、`NaN`が0に置き変わります。

In [117]:
df.fillna(0)

0         1         2         3
0  0.409686  0.127332  0.232334  0.472368
1  0.000000  0.308719  0.099518  0.838625
2  0.735781  0.911987  0.000000  0.982552
3  0.333130  0.581156  0.911326  0.673805
4  0.900381  0.254568  0.969245  0.841306
5  0.697316  0.653956  0.577141  0.000000
6  0.212289  0.870528  0.917358  0.000000
7  0.574902  0.365432  0.778180  0.000000
8  0.802476  0.911588  0.448293  0.000000
9  0.752408  0.741778  0.980731  0.000000

#### 前の値で埋める
`ffill`メソッドを適用すると、直前の行の値で埋めることができます。具体的には、2行1列目（インデックス「1」/カラム「0」の値）は先ほど

`df.iloc[1,0] = NA`

で`NA`にしましたが、直前の1行1列目の値は、0.893145でしたので、この値で埋めることができます。この処理は金融の時系列データの処理などで使うことができ、便利です。

In [118]:
df.fillna(method = 'ffill')

0         1         2         3
0  0.409686  0.127332  0.232334  0.472368
1  0.409686  0.308719  0.099518  0.838625
2  0.735781  0.911987  0.099518  0.982552
3  0.333130  0.581156  0.911326  0.673805
4  0.900381  0.254568  0.969245  0.841306
5  0.697316  0.653956  0.577141  0.841306
6  0.212289  0.870528  0.917358  0.841306
7  0.574902  0.365432  0.778180  0.841306
8  0.802476  0.911588  0.448293  0.841306
9  0.752408  0.741778  0.980731  0.841306

他に、平均値で穴埋めする方法もあります。これを**平均値代入法**といい、`mean`メソッドを使います。なお、注意点として、時系列データを扱う際、この方法は未来情報を含むことがある（過去に欠損したデータを、未来のデータを使った平均値で埋める）ので、気を付けましょう。


In [119]:
# 各カラムの平均値(確認用)
df.mean()

0    0.602041
1    0.572704
2    0.657125
3    0.761731
dtype: float64

In [120]:
df.fillna(df.mean())

0         1         2         3
0  0.409686  0.127332  0.232334  0.472368
1  0.602041  0.308719  0.099518  0.838625
2  0.735781  0.911987  0.657125  0.982552
3  0.333130  0.581156  0.911326  0.673805
4  0.900381  0.254568  0.969245  0.841306
5  0.697316  0.653956  0.577141  0.761731
6  0.212289  0.870528  0.917358  0.761731
7  0.574902  0.365432  0.778180  0.761731
8  0.802476  0.911588  0.448293  0.761731
9  0.752408  0.741778  0.980731  0.761731

他にも色々とオプションがあるので、`?df.fillna`等で調べてみてください。

欠損データについて、ここではサンプルデータにおいて、一定の値を機械的に置換しました。ただし、これらの方法はいつも使えるというわけではありません。データの状況、背景等を考え、適切に対処することが重要です。

#### <練習問題 4-13>

以下のデータに対して、1列でも`NaN`がある場合は削除し、その結果を表示してください。

In [121]:
# データの準備
import numpy as np
from numpy import nan as NA
import pandas as pd


df2 = pd.DataFrame(np.random.rand(15,6))

# NAにする
df2.iloc[2,0] = NA
df2.iloc[5:8,2] = NA
df2.iloc[7:9,3] = NA
df2.iloc[10,5] = NA


df2

0         1         2         3         4         5
0   0.161581  0.557745  0.180811  0.365589  0.025437  0.548313
1   0.421078  0.294969  0.046753  0.495783  0.430678  0.080125
2        NaN  0.224885  0.795201  0.693829  0.693913  0.587302
3   0.817315  0.402466  0.380375  0.155695  0.369437  0.733642
4   0.247733  0.160585  0.714020  0.080694  0.170367  0.157042
5   0.470570  0.090347       NaN  0.730314  0.475707  0.619626
6   0.349153  0.792649       NaN  0.410058  0.752320  0.495038
7   0.924592  0.334607       NaN       NaN  0.263934  0.616917
8   0.970935  0.194755  0.322675       NaN  0.786241  0.741476
9   0.323939  0.795866  0.148578  0.646092  0.130089  0.418628
10  0.095531  0.894436  0.584434  0.701178  0.440027       NaN
11  0.296139  0.206480  0.251601  0.250123  0.508324  0.923676
12  0.578330  0.512098  0.258208  0.773533  0.911133  0.215816
13  0.540076  0.628946  0.423177  0.297281  0.896136  0.786518
14  0.477888  0.313208  0.052573  0.609561  0.875702  0.421168

In [122]:
# write me!

#### <練習問題 4-14>

<練習問題 4-13>で準備したデータに対して、`NaN`を0で埋めてください。

In [123]:
# write me!

#### <練習問題 4-15>

<練習問題 4-13>で準備したデータに対して、`NaN`をそれぞれの列の平均値で埋めてください。

In [124]:
# write me!

### 4.5.2 異常データの扱い方
キーワード：異常値、箱ひげ図、パーセンタイル、VaR（Value At Risk）

次は、異常値（外れ値）についてです。異常値データの扱いは、そのままにして何もしないのか、異常値を除去するか、もっともらしい値に入れかえて使うかが問題になります。
そもそも異常値とは一体何でしょうか。実は、統一的な見解というものはなく、そのデータを扱うアナリストや意思決定者が判断することもあります。ビジネスの現場では、不正アクセスのパターン（セキュリティ分野）や機械の故障、金融リスク管理（VaR）など、さまざまな分野で使われており、それぞれ色々な方法でアプローチされています。

常値検出のアプローチには、単純に箱ひげ図などを書いて、あるパーセンタイル以上のデータを異常値としてみなす方法、正規分布を利用する方法、データの空間的な近さに基づく方法などがあります。他には以降の章で学ぶ機械学習（教師なし学習も含む）を用いた方法もあります。

ここでは特に練習問題はありませんが、興味のある方はぜひ巻末の参考文献「A-13」や参考URL「B-22」などで学んでください。

また、異常値の分野に関連して、極端な値を研究する極値統計学という分野もあります。データの中で大きな値をとる極値データの挙動について、さまざまな研究がなされており、稀ではありますがそれが起きれば非常に大きな影響を及ぼす現象（自然現象、災害など）を研究します。気象学だけではなく、ファイナンスや情報通信の分野でも応用されているので、興味のある方は参考文献「A-14」などを参照してみてください。

以上で、欠損値と異常値の扱いについてはこれで終わりになります。データ分析において、データの前処理が8割だとよく言われ、欠損データや異常値データには、たびたび遭遇します。また、世の中には実にさまざまな形式のデータが存在し、それらを整えるだけでも大変な作業です。ここで紹介したテクニックも重要ですが、それらに対してどのように対処していくのか戦略を立てることも重要です。参考文献「A-15」にも、ぜひ目を通してみてください。

***

## 4.6 時系列データの取り扱いの基礎

最後にPandasを使った時系列データの取り扱いについて学びます。ここでは、サンプルとして為替の時系列データを扱います。
あらかじめAppendixを参考に`pandas-datareader`というライブラリをダウンロードしてインストールしてから進めてください。

インストールしたら、次のようにインポートしてください。

In [125]:
import pandas_datareader.data as pdr

### 4.6.1 時系列データの処理と変換
キーワード：リサンプリング、シフト

ここでは、サンプルデータに含まれる2001/1/2から2016/12/30までのドル円の為替レートデータ（DEXJPUS）を使います。日ごとのレートデータで、欠損している日（休日など）もあります。

In [126]:
start_date = '2001/1/2'
end_date = '2016/12/30'

fx_jpusdata = pdr.DataReader('DEXJPUS', 'fred', start_date, end_date) 

`head`メソッドを使って、読み込んだfx_jpusdataの先頭5行を読み出します。

In [127]:
fx_jpusdata.head()

DEXJPUS
DATE               
2001-01-02   114.73
2001-01-03   114.26
2001-01-04   115.47
2001-01-05   116.19
2001-01-08   115.97

サンプルには、15年分のデータがありますが、これをどう分析するかはそのビジネスニーズ次第です。たとえば、最後の2016年の4月のデータだけ欲しいこともありますし、月末のレートだけを見たいこともあります。さらに、上記では、2001/1/6はデータとしてありませんが、それを前日の値で埋めたいこともありますし、前の日と比べてどれだけレートが上がったのか調べたい場合もあるでしょう。これらのことはすべてPandasで簡単に計算することができます。

#### 特定の年月のデータを参照する

まずは、特定の年月のデータを参照する方法です。2016年の4月のデータだけ見たい場合は、以下のように年月を指定します。

In [128]:
fx_jpusdata['2016-04']

DEXJPUS
DATE               
2016-04-01   112.06
2016-04-04   111.18
2016-04-05   110.26
2016-04-06   109.63
2016-04-07   107.98
2016-04-08   108.36
2016-04-11   107.96
2016-04-12   108.54
2016-04-13   109.21
2016-04-14   109.20
2016-04-15   108.76
2016-04-18   108.85
2016-04-19   109.16
2016-04-20   109.51
2016-04-21   109.41
2016-04-22   111.50
2016-04-25   111.08
2016-04-26   111.23
2016-04-27   111.26
2016-04-28   108.55
2016-04-29   106.90

そのほか、特定の年や日にちにだけ抽出することもできます。
次に、月末レートだけ取り出してみましょう。`resample`メソッドの引数に`M`を指定することで、月ごとのデータを取り出し、`last`メソッドで末尾のデータを取り出しています。具体的には、以下の結果をみるとわかる通り、1月、2月、3月…の月末のレートを取り出せます。

In [129]:
fx_jpusdata.resample('M').last().head()

DEXJPUS
DATE               
2001-01-31   116.39
2001-02-28   117.28
2001-03-31   125.54
2001-04-30   123.57
2001-05-31   118.88

日付を取り出したい場合は「`D`」、年を取り出したい場合は「`Y`」を、それぞれ引数に指定します。このように、ある頻度のデータを、別の頻度のデータで取り出し直す処理を**リサンプリング**といいます。また、最後のデータではなく、その平均を計算したい場合は`mean`メソッドを使うことで計算できます。他にもいろいろとパラメータを設定できるので、必要な処理があるときに、調べてみてください。

#### 欠損がある場合の操作

次に、時系列データに欠損がある場合の処理をみていきます。欠損処理については、前の節でも扱った通り、さまざまな方法があります。先ほどのレートでは、2001/1/6がまずレコードとして存在していませんでしたが、日ごとにデータを用意したいときは、先ほどのリサンプリングを行います。具体的には、以下のようにします。

In [130]:
fx_jpusdata.resample('D').last().head()

DEXJPUS
DATE               
2001-01-02   114.73
2001-01-03   114.26
2001-01-04   115.47
2001-01-05   116.19
2001-01-06      NaN

上記より、2001/1/6は空のままなので、前の日の値で埋める処理をします。ここでは、次に示すようにffillメソッドを使います。



In [131]:
fx_jpusdata.resample('D').ffill().head()

DEXJPUS
DATE               
2001-01-02   114.73
2001-01-03   114.26
2001-01-04   115.47
2001-01-05   116.19
2001-01-06   116.19

#### データをズラして比率を計算する

次に、前日とのレート比較をしたい場合を考えます。上のデータ例でいうと、2001-01-02のレートは114.73で、2001-01-03のレートは114.26になり、その比率を計算することもできますが、それをすべての日付について適応させる処理をします。shiftメソッドを使うことで、インデックスは固定したまま、そのデータだけをずらすことができます。以下はデータを1つあとにずらしており、2001-01-02のレートは114.73でしたが、2001-01-03のレートとして扱われるようになります。

In [132]:
fx_jpusdata.shift(1).head()

DEXJPUS
DATE               
2001-01-02      NaN
2001-01-03   114.73
2001-01-04   114.26
2001-01-05   115.47
2001-01-08   116.19

このように加工すると、前日のレートと当日のレートの比率を一気に算出することができます。これがPandasを使うメリットです。なお、以下で2001-01-02が`NaN`になっているのは、その前日のデータがもともとないためです。

In [133]:
fx_jpusdata_ratio = fx_jpusdata / fx_jpusdata.shift(1)
fx_jpusdata_ratio.head()

DEXJPUS
DATE                
2001-01-02       NaN
2001-01-03  0.995903
2001-01-04  1.010590
2001-01-05  1.006235
2001-01-08  0.998107

なお、差分や比率を取る方法については、`diff`や`pct_change`などもありますので、興味がある方は調べてみてください。

#### <練習問題 4-16>

上記のようにして読み込んだfx_jpusdataを使って、年ごとの平均値の推移データを作成してください。

In [134]:
# write me!

### 4.6.2 移動平均
次に、時系列のデータ処理でよく使われる移動平均の処理方法をみていきます。さきほど扱った`fx_jpusdata`のデータについて、3日間の移動平均線を作成することを考えます。まず先頭から5行のデータを取り出してみます。

In [135]:
fx_jpusdata.head()

DEXJPUS
DATE               
2001-01-02   114.73
2001-01-03   114.26
2001-01-04   115.47
2001-01-05   116.19
2001-01-08   115.97

結果を見るとわかるように、2001-01-04までのデータは、2001-01-02が114.73、2001-01-03が114.26、2001-01-04が115.47ですから、その平均を計算すると114.82です。同様にして、2001-01-05、2001-01-06と続けて計算をしていきます。それにはPpandasの`rolling`メソッドを使うと、簡単に計算できます。以下は、その3日間の移動平均を計算した結果です。`rolling`メソッドを実行した後に、`mean`メソッドを使って平均を計算しています。

In [136]:
fx_jpusdata.rolling(3).mean().head()

DEXJPUS
DATE                  
2001-01-02         NaN
2001-01-03         NaN
2001-01-04  114.820000
2001-01-05  115.306667
2001-01-08  115.876667

移動平均ではなく標準偏差の推移を算出したいのなら、`mean`メソッドの代わりに`std`メソッドを使います。以下は3日間の標準偏差の推移です。

In [137]:
fx_jpusdata.rolling(3).std().head()

DEXJPUS
DATE                
2001-01-02       NaN
2001-01-03       NaN
2001-01-04  0.610000
2001-01-05  0.975312
2001-01-08  0.368963

`rolling`メソッドには、パラメータが他にもいろいろとありますので、必要に応じて調べて実行してみてください。

以上で、Pandasの章は終了です。一部、なかなかイメージを掴みにくい箇所もあったかもしれません。しかし、実際に「こんな感じでデータ加工や変換したいのになあ」と思ったときに、ここを参考にしてプログラミングをしてみてください。データ加工処理のニーズが出てきて、実際に使うことで一層理解が進む箇所かもしれません。ここで紹介したテクニックはほんの一部です。この他にも、さまざまなデータ処理・加工方法があるので、参考文献「A-10」などを読んで、手を動かして実行してみてください。

#### <練習問題 4-17>

<練習問題 4-16>で使用したfx_jpusdataを使って、20日間の移動平均データを作成してください。ただし`NaN`は削除してください。なお、レコードとして存在しないデータであれば、特に補填する必要はありません。

In [138]:
# write me!

***

## 4.7 総合問題

### ■総合問題4-1 データ操作

数学の成績を示すデータである「student-mat.csv」を使って、以下の問いに答えてください。

(1) 上記のデータに対して、年齢（`age`）×性別（`sex`）で`G1`の平均点を算出し、縦軸が年齢（`age`）、横軸が性別（`sex`）となるような表（テーブル）を作成しましょう。

(2) (1)で表示した結果テーブルについて、NaNになっている行（レコード）をすべて削除した結果を表示しましょう。

In [139]:
# write me!